br/pkg/lightning/backend/local/local.go

```diff
@@ -1630,43 +1293,76 @@ func (local *local) ImportEngine(ctx context.Context, engineUUID uuid.UUID, regi
 
 	failpoint.Inject("ReadyForImportEngine", func() {})
 
-	for {
-		unfinishedRanges := lf.unfinishedRanges(ranges)
-		if len(unfinishedRanges) == 0 {
-			break
+	var (
+		workGroup, workerCtx = errgroup.WithContext(ctx)
+		// jobCh is unbuffered so when we finished sending all jobs, we can make sure all jobs have been
+		// received by workers.
+		jobCh = make(chan *regionJob)
+		// jobWg is the number of jobs that need to be processed by worker in this round.
+		jobWg           sync.WaitGroup
+		needRetryJobs   []*regionJob
+		needRetryJobsMu sync.Mutex
+		putBackFn       = func(job *regionJob) {
+			needRetryJobsMu.Lock()
+			defer needRetryJobsMu.Unlock()
+			needRetryJobs = append(needRetryJobs, job)
 		}
-		log.FromContext(ctx).Info("import engine unfinished ranges", zap.Int("count", len(unfinishedRanges)))
+	)
 
-		// if all the kv can fit in one region, skip split regions. TiDB will split one region for
-		// the table when table is created.
-		needSplit := len(unfinishedRanges) > 1 || lfTotalSize > regionSplitSize || lfLength > regionSplitKeys
-		// split region by given ranges
-		for i := 0; i < maxRetryTimes; i++ {
-			err = local.SplitAndScatterRegionInBatches(ctx, unfinishedRanges, lf.tableInfo, needSplit, regionSplitSize, maxBatchSplitRanges)
-			if err == nil || common.IsContextCanceledError(err) {
-				break
-			}
+	for i := 0; i < local.workerConcurrency; i++ {
+		workGroup.Go(func() error {
+			return local.startWorker(workerCtx, jobCh, &jobWg, putBackFn)
```

Comment:
It's not clear to me when job is done and how job is put back. I think a more straightforward way is run the job and return whether the job needs to retry and finished or unfinished ranges.

Reply:
Not understand, "return whether the job needs..." means the goroutine is exited. If you mean add retry and other properties to result structure it's also duplicated, because all these information can be found in job.

Reply:
What I mean is to control `jobWg` and `putBackFn` here directly instead of passing them to `startWorker`.